python/lsst/daf/butler/core/quantum.py

```diff
@@ -569,6 +569,87 @@ def _reduceFactory(
             datastore_records=datastore_records,
         )
 
+    def update_output_run(self, run: str, dataset_id_map: MutableMapping[DatasetId, DatasetId]) -> Quantum:
+        """Update name of the run collection in each output dataset reference
+        in this Quantum.
+
+        Parameters
+        ----------
+        run : `str`
+            Name of the run collection for all output datasets.
+        dataset_id_map : `~collections.abc.MutableMapping`
+            This mapping is populated with correspondence of the old dataset
+            IDs to new dataset IDs for every updated `DatasetRef`.
+
+        Notes
+        -----
+        This method supports updating of output run at the level of whole
+        quantum graph. To correctly update dataset IDs of intermediate datasets
+        in a graph the `update_output_run` has to run first on all quanta in a
+        graph. ``dataset_id_map`` should also be populated with the updated
+        dataset IDs of initOutputs (which are graph property). When all output
+        refs are updated, the `update_input_run` has to be executed for each
+        quantum.
+        """
+
+        def _update_output_refs(
+            refs: Iterable[DatasetRef], run: str, dataset_id_map: MutableMapping[DatasetId, DatasetId]
+        ) -> Iterator[DatasetRef]:
+            for ref in refs:
+                new_ref = ref.replace(run=run)
+                dataset_id_map[ref.id] = new_ref.id
+                yield new_ref
+
+        outputs = {
+            dataset_type: tuple(_update_output_refs(refs, run, dataset_id_map))
+            for dataset_type, refs in self._outputs.items()
+        }
+
+        return Quantum(
+            taskName=self._taskName,
+            dataId=self._dataId,
+            initInputs=self._initInputs,
+            inputs=self._inputs,
+            outputs=outputs,
+            datastore_records=self._datastore_records,
+        )
+
+    def update_input_run(self, run: str, dataset_id_map: Mapping[DatasetId, DatasetId]) -> Quantum:
+        """Update name of the run collection in input dataset references in
+        this Quantum.
+
+        Parameters
+        ----------
+        run : `str`
+            Name of the run collection for all output datasets.
+        dataset_id_map : `~collections.abc.MutableMapping`
+            The mapping with correspondence of the old dataset IDs to new
+            dataset IDs, filled previously by `update_output_run`.
+        """
+
+        def _update_input_refs(
+            refs: Iterable[DatasetRef], run: str, dataset_id_map: Mapping[DatasetId, DatasetId]
+        ) -> Iterator[DatasetRef]:
+            for ref in refs:
+                if dataset_id := dataset_id_map.get(ref.id):
```

Comment:
Add comment that this is for intermediates only and if it's not already in dataset_id_map the ref won't be modified?